### PR TITLE
feat(startup): update logic for metadb update on startup, skip unmodi…

### PR DESCRIPTION
**Pull request by laurentiuNiculae**
…fied repos

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/meta/boltdb/boltdb.go [hunk 1]
```diff
@@ -237,6 +295,17 @@ func (bdw *BoltDB) SetRepoReference(ctx context.Context, repo string, reference
 	return err
 }
 
+func updatedRepoLastUpdated(repo string, repoLastUpdatedBuck *bbolt.Bucket) error {
```
Comment (peusebiu) [inline pkg/meta/boltdb/boltdb.go:298]: updateRepoLastUpdated instead?

you can also make protoTime an arg, and also use this function in SetRepoMeta()

File: pkg/storage/imagestore/imagestore.go [hunk 2]
```diff
@@ -1505,6 +1505,25 @@ func (is *ImageStore) GetIndexContent(repo string) ([]byte, error) {
 	return buf, nil
 }
 
+func (is *ImageStore) StatIndex(repo string) (bool, int64, time.Time, error) {
+	dir := path.Join(is.rootDir, repo, "index.json")
```
Comment (peusebiu) [inline pkg/storage/imagestore/imagestore.go:1509]: is not a dir?

**Review by peusebiu — APPROVED** [review]
lgtm

**Review by andaaron — COMMENTED** [review]
Overall the approach looks good, but what happens if the contents of the pre-existing buckets does not match the current data?

Let's say for example you do a zot upgrade to a binary built from this PR and the repos in the DB do not have the last updated data/buckets?

**Comment by rchincha** [comment]
@laurentiuNiculae can you summarize your changes/thoughts in this PR ... ideally in the commit msg itself.

File: pkg/meta/parse.go [hunk 3]
```diff
@@ -27,21 +27,49 @@ const (
...
 
 		return err
 	}
 
-	for i, repo := range allRepos {
-		log.Info().Int("total", len(allRepos)).Int("progress", i).Str("current-repo", repo).
+	for _, repo := range getReposToBeDeleted(allStorageRepos, allMetaDBRepos) {
+		err := metaDB.DeleteRepoMeta(repo)
+		if err != nil {
+			log.Error().Err(err).Str("rootDir", rootDir).
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/meta/parse.go]: Updated

File: pkg/storage/imagestore/imagestore.go [hunk 4]
```diff
@@ -1505,6 +1505,25 @@ func (is *ImageStore) GetIndexContent(repo string) ([]byte, error) {
...
+
+	fileInfo, err := is.storeDriver.Stat(repoIndexPath)
+	if err != nil {
+		if errors.As(err, &driver.PathNotFoundError{}) {
+			is.log.Error().Err(err).Str("indexFile", repoIndexPath).Msg("index.json doesn't exist")
+
+			return false, 0, time.Time{}, zerr.ErrRepoNotFound
+		}
+
+		is.log.Error().Err(err).Str("indexFile", repoIndexPath).Msg("failed to read index.json")
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/storage/imagestore/imagestore.go:1519]: I tried, but I don't know how. Stat can't be made to fail by changing the permissions

File: pkg/test/oci-utils/store.go [hunk 5]
```diff
@@ -23,7 +23,7 @@ func GetDefaultImageStore(rootDir string, log zLog.Logger) stypes.ImageStore {
 	)
 }
 
-func GetDefaultStoreController(rootDir string, log zLog.Logger) stypes.StoreController {
+func GetDefaultStoreController(rootDir string, log zLog.Logger) storage.StoreController {
```
[the remark this answers is not in the document]
Comment (laurentiuNiculae) [reply pkg/test/oci-utils/store.go:26]: I have kept the interface type

**Comment by laurentiuNiculae** [comment]
> @laurentiuNiculae can you summarize your changes/thoughts in this PR ... ideally in the commit msg itself.

Now we parse just repos who were updated after the server is closed. 

To do this we:
 - Store the timestamp when the repo was last modified in MetaDB, I'll note it `DBLastUpdated`
 - We compare that timestamp with the current "LastModified" time given by the storage (1). I'll note it `StorageLastUpdated`
- If `StorageLastUpdated` is after `DBLastUpdated` we must parse this repo and update MetaDB, otherwise we can just skip it.

(1) LastModified refers to the last modified value of the index.json file for the repo

**Comment by rchincha** [comment]
Does boltdb do internal checksums to detect data corruptions?
If corrupted, repair.

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm
